File: solutions/elixir/046-subtree-of-another-tree.ex
# Problem 46: Subtree of Another Tree (LeetCode 572)
# Difficulty: Easy
# Language: Elixir
# 
defmodule SubtreeCheck do
  def is_subtree(nil, nil), do: true
  def is_subtree(nil, _sub), do: false
  def is_subtree(root, sub) do
    same_tree?(root, sub) or
      is_subtree(root.left, sub) or
      is_subtree(root.right, sub)
  end

  defp same_tree?(nil, nil), do: true
  defp same_tree?(nil, _), do: false
  defp same_tree?(_, nil), do: false
  defp same_tree?(a, b) do
    a.val == b.val and same_tree?(a.left, b.left) and same_tree?(a.right, b.right)
  end
end

File: solutions/clojure/046-subtree-of-another-tree.clj
;; Problem 46: Subtree of Another Tree (LeetCode 572)
;; Difficulty: Easy
;; Language: Clojure
;; 
(defn same-tree? [a b]
  (cond
    (and (nil? a) (nil? b)) true
    (or (nil? a) (nil? b)) false
    :else (and (= (:val a) (:val b))
               (same-tree? (:left a) (:left b))
               (same-tree? (:right a) (:right b)))))

(defn is-subtree [root sub]
  (cond
    (nil? root) (nil? sub)
    :else (or (same-tree? root sub)
              (is-subtree (:left root) sub)
              (is-subtree (:right root) sub))))
